Задание II

PrideSt opened the pull request:
В пакете internal/utils вашего репозитория нужно реализовать экспортируемые функции для

- [x] Разделение на слайса на батчи - исходный слайс конвертировать в слайс слайсов - чанки одинкового размера (кроме последнего)
- [x] Обратный ключ - происходит конвертация отображения (“ключ-значение“) в отображение (“значение-ключ“)
- [x] Фильтрация по захордкоженному списку - нужно отфильтровать входной слайс по критерию отсутствия элемента в захардкоженном списке
- [x] В качестве типов сейчас будем использовать встроенные типы: целые числа или строковые

@see https://golangschoolozon.slack.com/archives/C029HUD82SF/p1628190396013300

Commit:
issue-2 add utils package

add functions ChunkSlice, FlipMap, FilterByValues

Commit:
issue-2 remove comment with too long string

dmzkrsk commented:
Если сразу оба MR кидаешь, то у второго сделай diff, относительно первого, иначе получается, что дважды придется смотреть первый MR

Commit:
Merge branch 'issue-1' into issue-2

PrideSt commented:
Вот диф: https://github.com/ozonva/ova-joke-api/compare/issue-1...issue-2
В дальнейшем я буду стараться от мастера делать задачи, но не хотелось еще раз линтер настраивать, поэтому отбранчевался от первой ветки

Commit:
issue-2 revert FilterByValues from white to black list, add some tests

Commit:
issue-2 add some test cases

Commit:
Merge branch 'main' into issue-2

PrideSt commented:
Выяснилось что для required апрувов нужны апрувы пользователей с write правами на репозиторий, выдал права.
https://github.com/ozonva/ova-joke-api/invitations
Подмержил 1-ю часть в мастер, обновил этот ПР чтобы он показывал только новые комиты.
Упустил момент что compare веток не позволяет оставлять комментарии. Редко пользуюсь github не знаю некоторых тонкостей. Для проверки верну настройки для protected branch требование на 1 апрув от пользователя с write правами.

dmzkrsk commented on internal/utils/utils.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	chunksCnt := len(data) / sz
+	if len(data)%sz > 0 {
```
Можно использовать [integer ceil division](https://stackoverflow.com/questions/2745074/fast-ceiling-of-an-integer-division-in-c-c/2745086#2745086)

PrideSt replied:
Изучил способ, поправил

dmzkrsk commented on internal/utils/utils.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	result := make([][]string, 0, chunksCnt)
+	for i := 0; i*sz < len(data); i++ {
```
Сложно читать, когда на каждой строчке умножение, причем одинаковое (`i*sz`)

Почему бы не сделать так: `for i := 0; i < len(data); i += sz {`

PrideSt replied:
Поправил

dmzkrsk commented on internal/utils/utils.go:
```diff
@@ -0,0 +1,82 @@
+	for i := 0; i*sz < len(data); i++ {
+		chunk := make([]string, minInt(sz, len(data)-i*sz))
+		copy(chunk, data[i*sz:minInt((i+1)*sz, len(data))])
+		result = append(result, chunk)
```
`make`+`copy` можно выкинуть и сделать просто `result = append(result, data[x:y])`

PrideSt replied:
Поправил.
Я копировал преднамеренно, идея была в том, чтобы возвращать новый слайс с чанками никак не связанными со старым и защитить тем самым их друг от друга в плане модификации и переалокаций. В постановке задачи не описано желаемое поведение. Я подумал, что перекомпановка исходного слайса не очень полезна т.к. можно использовать преобразование двумерных индексов в одномерные по формуле n = y * width + x

Commit:
issue-2 fix reviewer's comments

fix simplify ChunkSlice method
fix handle division by zero error
brk ChunkSlice returns references on given slice not copy
